Adds configuration option to set max_fast_inline

fdel15 opened the pull request:
### Description
Closes #2361

Adds a config option to set `max_fast_inline` and override the `MAX_FAST_INLINE` default. You can set this variable to `Float::Infinity` to address use cases that wish to maintain the pre version 3.12.2 behavior (as discussed in the issue).

~Adds an ENV option to set MAX_FAST_INLINE constant. You can set this variable to zero to address use cases that wish to maintain the pre version 3.12.2 behavior (as discussed in the issue).~

~I didn't add any tests because it is not clear to me where to add them. This change affects the `Puma::Server#process_client` method, and it doesn't look like we have any tests set up at the moment. I also did not see any tests written when this change was introduced in the commits referenced in the issue https://github.com/puma/puma/compare/v3.12.1...v3.12.2~

### Your checklist for this pull request
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [x] I have added an entry to [History.md](../blob/master/History.md) if this PR fixes a bug or adds a feature. If it doesn't need an entry to HISTORY.md, I have added `[changelog skip]` or `[ci skip]` to the pull request title.
- [x] I have added appropriate tests if this PR fixes a bug or adds a feature.
- [x] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [ ] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [x] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed, including Rubocop.

nateberkopec commented on History.md:
```diff
@@ -9,6 +9,11 @@
 * Refactor
   * Consolidate option handling in Server, Server small refactors, doc chang (#2389)
 
+## 5.0.3 / 2020-10-02
```
Please add any changelog entries to the "master" entry, above this one

nateberkopec commented on lib/puma/const.rb:
```diff
@@ -124,7 +124,7 @@ module Const
     # but very "greedy" client across 10 requests. This prevents a not
     # well behaved client from monopolizing the thread forever.
-    MAX_FAST_INLINE = 10
+    MAX_FAST_INLINE = ENV.fetch('MAX_FAST_INLINE', 10).to_i
```
Rather than have a constant set by an ENV variable, I think we should add this to the configuration object and then read it from there.

nateberkopec replied:
I think it would be cleaner to make this an attribute of the Server object, which is read from configuration. 

We don't have any env variables setting constants right now, probably shouldn't start.

nateberkopec commented on lib/puma/server.rb:
```diff
@@ -426,7 +426,7 @@ def process_client(client, buffer)
             check_for_more_data = @status == :run
 
-            if requests >= MAX_FAST_INLINE
+            if requests >= MAX_FAST_INLINE && MAX_FAST_INLINE.positive?
```
e.g. server should have `@max_fast_inline` set by the options, and then that is read here. Don't check here to see if it's positive, check when you're setting it.

nateberkopec commented on History.md:
```diff
@@ -9,6 +9,11 @@
 
+## 5.0.3 / 2020-10-02
+
+* Features 
```
Please add changelog entries to the "master" entry above this one

nateberkopec commented:
Cheers @fdel15! Thanks for your contribution!

Let's move this into the configuration object, like the other attributes on Server which get set as "options".

fdel15 commented:
Thanks for the feedback Nate. I should have time this weekend to make the requested changes.

fdel15 commented:
>  You can set this variable to zero to address use cases that wish to maintain the pre version 3.12.2 behavior

I originally had it working this way because an ENV variable is a string, so the INFINITY option that the issue suggested would not work. Since we are no longer using an ENV variable, we can allow users to set configure `max_fast_inline` to INFINITY to maintain the pre version 3.12.2 behavior.

dentarg commented:
@fdel15 can you update the title and PR description here to reflect what is actually happening now?

Could be discussed elsewhere, but I'll raise it now because ([the `History.md` diff here looks a bit funny](https://github.com/puma/puma/pull/2406/files#diff-e1bbd4f15e3b63427b4261e05b948ea8)). @nateberkopec @MSP-Greg should we remove the requirement that contributors update the changelog? now that @MSP-Greg has some tooling to do it

dentarg commented:
Also: I don't think CI should be skipped for this PR

fdel15 commented:
> can you update the title and PR description here to reflect what is actually happening now?

Updated.

> Also: I don't think CI should be skipped for this PR

I removed the CI skip. I added it because I didn't add any tests in the PR. I feel like now that I am doing more than just adding an ENV variable I should be adding tests, but I am not sure how to test this feature. Any guidance on this would be greatly appreciated.

MSP-Greg commented:
@fdel15

Thanks for the PR.  This jumps around, so I won't inline it.  In server.rb
```ruby
attr_reader :max_fast_inline
```
and:
```ruby
if requests >= max_fast_inline
```

One would normally prefer that the API only exposes methods and attributes that are required externally. `max_fast_inline`  is currently not used anywhere other than the `Server` object, so it shouldn't have a reader.  The conditional can just use the instance variable set in the `initialize` method.

Re History.md, you may need to rebase for that.  Also, the code I've got for it only adds the link info.  At present it doesn't add items to the log, as that would require exact tagging for what 'group' an entry is listed in.

nateberkopec commented:
Agree with @MSP-Greg's feedback that we want to keep the interface private.

MSP-Greg commented:
This needs to be added to `puma/dsl.rb`?

fdel15 commented:
> One would normally prefer that the API only exposes methods and attributes that are required externally.

Feedback addressed. Good call about not exposing an internal attribute.

> This needs to be added to puma/dsl.rb?

I think this is a good call as well. I added a `max_fast_inline` method and a test in `test/config`

MSP-Greg commented:
@fedl15

Do you see the message 'This branch cannot be rebased due to conflicts'?   You may need to remove the commits in your PR branch that are duplicated, update your forks' master branch, then rebase on master in your PR branch.  I think that made sense.  I can help with more info if needed.

You'll notice all the files listed as changed above.  It makes it hard to see what's changed when the master branch's commits are mixed in...

fdel15 commented:
> Do you see the message 'This branch cannot be rebased due to conflicts'?

I thought I fixed the conflicts when I rebased. I must've done something silly though that caused all the files to be listed as changed. Sorry for not catching that earlier. I cherry picked all my commits for the branch and force pushed them up. This should be good for review now.

dentarg commented on History.md:
```diff
@@ -1,7 +1,18 @@
 ## Master
 
-* Features
-  * Your feature goes here <Most recent on the top, like GitHub> (#Github Number)
```
I think we want to keep this line :)

dentarg commented on History.md:
```diff
@@ -1,7 +1,18 @@
+  * client.rb - remove JRuby specific 'finish' code (#2412)
+  * Consolidate fast_write calls in Server, extract early_hints assembly (#2405)
+  * Remove upstart from docs (#2408)
+  * Consolidate option handling in Server, Server small refactors, doc changes (#2389)
```
These lines should not be here, they are already in `History.md` (https://github.com/puma/puma/pull/2406/commits/72ed95944cd7f848413f021684be5c0fa114ed9b adds them for some reason)

fdel15 replied:
This should be good now

nateberkopec commented on lib/puma/dsl.rb:
```diff
@@ -810,5 +810,12 @@ def fork_worker(after_requests=1000)
+    # the reactor to be subject to normal ordering.
+    #
+    def max_fast_inline(num_of_requests)
+      @options[:max_fast_inline] = Float(num_of_requests)
```
I don't think it really matters, but why a float here? We can't have a non-integer number of requests, can we?

fdel15 replied:
To handle the use case mentioned in the issue to be able to set the value to Infinity. Using Float seemed like the simplest solution to me.

nateberkopec replied:
Got it!

nateberkopec commented:
Solid. Thanks for seeing this one through!

nateberkopec commented:
Whoops. I kinda forgot I wanted the next release to be a patchlevel. I may revert and then quickly re-revert this on Monday when we release.
